lib/Conversion/CGGIToJaxite/CGGIToJaxite.cpp

```diff
@@ -0,0 +1,306 @@
+}
+
+/// Returns the Value corresponding to a server key in the FuncOp containing
+/// this op.
+FailureOr<Value> getContextualJaxiteServerKeySet(Operation *op) {
+  int num_inputs = op->getParentOfType<func::FuncOp>()
+                       .getBody()
+                       .getBlocks()
+                       .front()
+                       .getNumArguments();
+  if (num_inputs < 2) {
+    return op->emitOpError() << "Found less than 2 ops in function input. Did "
+                                "the AddServerKeyArg pattern fail to run?";
+  }
+  auto block_args_list = op->getParentOfType<func::FuncOp>()
+                             .getBody()
+                             .getBlocks()
+                             .front()
+                             .getArguments();
+  auto it = block_args_list.begin();
+  while (!mlir::isa<jaxite::ServerKeySetType>(it->getType())) {
+    it++;
+  }
+  if (it == block_args_list.end() ||
+      !mlir::isa<jaxite::ServerKeySetType>(it->getType())) {
+    return op->emitOpError() << "Cannot find Params argument. Did the "
+                                "AddServerKeyArg pattern fail to run?";
+  }
+  return *it;
+}
+
+/// Returns the Value corresponding to params in the FuncOp containing
+/// this op.
+FailureOr<Value> getContextualJaxiteParams(Operation *op) {
+  int num_inputs = op->getParentOfType<func::FuncOp>()
+                       .getBody()
+                       .getBlocks()
+                       .front()
+                       .getNumArguments();
+  if (num_inputs < 2) {
+    return op->emitOpError() << "Found less than 2 ops in function input. Did "
+                                "the AddServerKeyArg pattern fail to run?";
+  }
+  auto block_args_list = op->getParentOfType<func::FuncOp>()
+                             .getBody()
+                             .getBlocks()
+                             .front()
+                             .getArguments();
+  auto it = block_args_list.begin();
+  while (!mlir::isa<jaxite::ParamsType>(it->getType())) {
+    it++;
+  }
+  if (it == block_args_list.end() ||
+      !mlir::isa<jaxite::ParamsType>(it->getType())) {
+    return op->emitOpError() << "Cannot find Params argument. Did the "
+                                "AddServerKeyArg pattern fail to run?";
+  }
+  return *it;
+}
+
+template <class Op>
+struct GenericOpPattern : public OpConversionPattern<Op> {
+  using OpConversionPattern<Op>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Op op, typename Op::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Type> retTypes;
+    if (failed(this->getTypeConverter()->convertTypes(op->getResultTypes(),
+                                                      retTypes)))
+      return failure();
+    rewriter.replaceOpWithNewOp<Op>(op, retTypes, adaptor.getOperands(),
+                                    op->getAttrs());
+
+    return success();
+  }
+};
```

Comment:
Didnt quite understand how to use ConvertAny pattern. Can you point me to an example usage?

Reply:
https://github.com/google/heir/blob/dd08bd681c97739a0208491cab3af296587632ee/lib/Dialect/TensorExt/Transforms/AlignTensorSizes.cpp#L173

Since it just replaces the op with type converted inputs and outputs, you can just add this pattern and ensure that the ops you want to target are marked as dynamically legal (which you've done)

Reply:
Then i guess i dont need the GenericOpPattern (removed it) and also do not need to explicitly add tensor::FromElementsOp, tensor::ExtractOp as dynamically legal. The tests seem to pass. So im resolving this.

Done.

Reply:
I was surprised you'd need to remove the callback. I patched and realized the reason why is because the jaxite dialect is using LWE types right now, and not jaxite types. So actually tensor::FromElementsOp and tensor::ExtractOp were already legal since they don't require any rewrite. If you migrate over to jaxite types you'll need to add a target callback to mark ops as dynamically legal only when the types are valid